### PR TITLE
[PC-6904] Codepush install during splashscreen

### PR DESCRIPTION
**Pull request by justinemignot**
Link to JIRA ticket: https://passculture.atlassian.net/browse/PC-XXX

## Checklist

I have:

- [x] Made sure the title of my PR follows the convention `[PC-XXX] <summary>`.
- [x] Made sure my feature is working on the relevant real / virtual devices.
- [x] Written **unit tests** for my feature.
- [x] Added new reusable components to the AppComponents page and communicate to the team on slack
- [x] Added a **screenshot** for UI tickets.
- [x] Attached a **ticket number** for any added TODO/FIXME \
       (for tech tasks, give the best context about the TODO resolution: what/who/when).

## Screenshots

| \*iOS - [Phone name] | \*Android - [Phone name] | Tablet - [Phone name] | Little - [Phone name] |
| -------------------- | :----------------------: | --------------------: | --------------------: |
|                      |                          |                       |                       |

## Deploy hard

If native code (ios/android) was modified, **after** the PR is merged, on the master branch, upgrade the **app version** (+1 patch):

- if you want an hard deployment of the testing environment, use `yarn trigger:testing:deploy`

### CODE REVIEW

File: src/features/navigation/RootNavigator/RootNavigator.tsx [hunk 1]
```diff
@@ -51,8 +53,11 @@ export const RootNavigator: React.FC = () => {
       analytics.logScreenView(routeName)
     }
   }
+  const { status } = useCodePush()
 
-  useHideSplashScreen({ shouldHideSplashScreen: !!initialRouteName })
+  useHideSplashScreen({
+    shouldHideSplashScreen: !!initialRouteName && status === CodePush.SyncStatus.UP_TO_DATE,
```
Comment (antoinewg) [inline src/features/navigation/RootNavigator/RootNavigator.tsx:59]: cool 👍

File: src/libs/codepush/CodePushProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import React, { createContext, useContext } from 'react'
+import CodePush from 'react-native-code-push' // @codepush
```
Comment (antoinewg) [inline src/libs/codepush/CodePushProvider.tsx:2]: c'est nécessaire les commentaires //@codepush ?

File: src/App.tsx [hunk 3]
```diff
@@ -63,37 +39,31 @@ const queryClient = new QueryClient({
...
+            </GeolocationWrapper>
+          </ErrorBoundary>
+        </QueryClientProvider>
+      </SafeAreaProvider>
+    </CodePushProvider>
   )
 }
 
-// @codepush
-const useCodePush = env.FEATURE_FLAG_CODE_PUSH && !__DEV__
```
Comment (antoinewg) [inline src/App.tsx:92]: cool d'enlever ça 👍

**Review by antoinewg — APPROVED** [review]

File: src/features/navigation/__tests__/RootNavigator.test.tsx [hunk 4]
```diff
@@ -32,6 +32,9 @@ jest.mock('features/firstLogin/achievements/pages/FirstTutorial/FirstTutorial',
     return <Text>FirstTutorial screen</Text>
   },
 }))
+jest.mock('libs/codepush/CodePushProvider', () => ({
+  useCodePush: jest.fn(() => ({ status: 'UP_TO_DATE' })),
```
Comment (BenStirrup) [inline src/features/navigation/__tests__/RootNavigator.test.tsx:36]: Je vais devoir te demander de mettre dans le test le cas où status n'est pas égale à 'UP_TO_DATE'^^

File: src/libs/codepush/CodePushProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  shouldWrapWithCodePush ? CodePush(codePushOptionsAuto)(AppComponent) : AppComponent
+
+// @ts-ignore no-param
+const CodePushContext = createContext<CodePushContext>({})
+
+export const useCodePush = () => useContext<CodePushContext>(CodePushContext)
+
+export const CodePushProvider = CodePushWrapper(
+  // @ts-ignore no-param
+  class App extends React.Component<Record<string, unknown>, CodePushContext> {
```
Comment (kopax-polyconseil) [inline src/libs/codepush/CodePushProvider.tsx:28]: C'est obligé ici d'utiliser une class plutôt qu'une function  ?

Comment (justinemignot) [reply src/libs/codepush/CodePushProvider.tsx:28]: oui pour utiliser `codePushStatusDidChange` c'est un peu ça le hack :)

**Review by BenStirrup — APPROVED** [review]

**Pull request merged**